src/ucs/sys/topo.c

```diff
@@ -19,9 +52,138 @@ ucs_sys_bus_id_t ucs_sys_bus_id_unknown = { .domain   = 0xffff,
                                             .function = 0xff
                                           };
 
+static ucs_bus_id_bit_rep_t ucs_topo_get_bus_id_bit_repr(const ucs_sys_bus_id_t *bus_id)
+{
+    ucs_bus_id_bit_rep_t bit_rep = 0;
+
+    bit_rep |= bus_id->domain;
+    bit_rep <<= 24;
+    bit_rep |= bus_id->bus;
+    bit_rep <<= 16;
+    bit_rep |= bus_id->slot;
+    bit_rep <<= 8;
+    bit_rep |= bus_id->function;
+
+    return bit_rep;
+}
+
+void ucs_topo_init()
+{
+    if (ucs_topo_initialized) return;
+
+    ucs_recursive_spinlock_init(&ucs_topo_sys_dev_to_bus_lock, 0);
+    ucs_recursive_spinlock_init(&ucs_topo_bus_to_sys_dev_lock, 0);
+    kh_init_inplace(bus_to_sys_dev, &ucs_topo_bus_to_sys_dev_hash);
+    kh_init_inplace(sys_dev_to_bus, &ucs_topo_sys_dev_to_bus_hash);
+
+    ucs_topo_initialized = 1;
+}
+
+void ucs_topo_cleanup()
+{
+    ucs_status_t status;
+    ucs_sys_device_t sys_dev;
+    ucs_sys_bus_id_t bus_id;
+
+    if (!ucs_topo_initialized) return;
+
+    kh_foreach_value(&ucs_topo_bus_to_sys_dev_hash, sys_dev, {
+        ucs_debug("sys_device_t %u not released", sys_dev);
+    })
+
+    kh_destroy_inplace(bus_to_sys_dev, &ucs_topo_bus_to_sys_dev_hash);
+
+    status = ucs_recursive_spinlock_destroy(&ucs_topo_bus_to_sys_dev_lock);
+    if (status != UCS_OK) {
+        ucs_warn("ucs_recursive_spinlock_destroy() failed: %s",
+                 ucs_status_string(status));
+    }
+
+    kh_foreach_value(&ucs_topo_sys_dev_to_bus_hash, bus_id, {
+        ucs_debug("bus id %ld not released",
+                   ucs_topo_get_bus_id_bit_repr(&bus_id));
+    })
+
+    kh_destroy_inplace(sys_dev_to_bus, &ucs_topo_sys_dev_to_bus_hash);
+
+    status = ucs_recursive_spinlock_destroy(&ucs_topo_sys_dev_to_bus_lock);
+    if (status != UCS_OK) {
+        ucs_warn("ucs_recursive_spinlock_destroy() failed: %s",
+                 ucs_status_string(status));
+    }
+
+    ucs_topo_initialized = 0;
+}
+
+static int ucs_topo_compare_bus_id(const ucs_sys_bus_id_t *bus_id1, 
+                                   const ucs_sys_bus_id_t *bus_id2)
+{
+    if (bus_id1->domain == bus_id2->domain &&
+        bus_id1->bus == bus_id2->bus &&
+        bus_id1->slot == bus_id2->slot &&
+        bus_id1->function == bus_id2->function) {
+        return 1;
+    } else {
+        return 0;
+    }
+}
+
 ucs_status_t ucs_topo_find_device_by_bus_id(const ucs_sys_bus_id_t *bus_id,
                                             ucs_sys_device_t *sys_dev)
 {
+    khiter_t hash_it;
+    //ucs_status_t status;
+    ucs_topo_kh_put_t kh_put_status;
+
+    if (ucs_topo_compare_bus_id(bus_id, &ucs_sys_bus_id_unknown)) {
+        *sys_dev = UINT_MAX;
+        ucs_debug("found unknown device index %u for bus id %ld",
+                  *sys_dev, ucs_topo_get_bus_id_bit_repr(bus_id));
+        return UCS_OK;
+    }
+    
+    *sys_dev = UINT_MAX;
```

Comment:
can do that once before line 138. Or move it to the relevant cases only (when put failed)